File: ruby/lib/nit/commands/passthrough.rb
# frozen_string_literal: true

module Nit
  module Commands
    class Passthrough
      # Format passthrough output - just show first non-empty line or "ok"
      def format(output)
        return error_line(output.stderr) unless output.success

        all_lines = output.stdout.lines + output.stderr.lines
        first_line = all_lines.find { |l| !l.strip.empty? }
        first_line&.strip || "ok"
      end

      # Build git command for passthrough (command and args passed as-is)
      def build_command(repo_path, args)
        raise ArgumentError, "No git command specified" if args.empty?

        command = args.first
        extra_args = args[1..] || []
        GitCommand.new(repo_path, command, extra_args)
      end

      private

      def error_line(stderr)
        line = stderr.lines.find { |l| !l.strip.empty? } || "unknown error"
        "ERROR: #{line.strip}"
      end
    end
  end
end

File: fit-rust/src/commands/passthrough.rs
use anyhow::Result;
use std::path::PathBuf;
use std::process::Output;

use crate::runner::{run_parallel, ExecutionContext, GitCommand, OutputFormatter};

struct PassthroughFormatter;

impl OutputFormatter for PassthroughFormatter {
    fn format(&self, output: &Output) -> String {
        let stdout = String::from_utf8_lossy(&output.stdout);
        let stderr = String::from_utf8_lossy(&output.stderr);

        if !output.status.success() {
            let error_line = stderr
                .lines()
                .find(|l| !l.trim().is_empty())
                .unwrap_or("unknown error");
            return format!("ERROR: {}", error_line);
        }

        // For passthrough, just show first non-empty line or "ok"
        stdout
            .lines()
            .chain(stderr.lines())
            .find(|l| !l.trim().is_empty())
            .unwrap_or("ok")
            .trim()
            .to_string()
    }
}

pub fn run(ctx: &ExecutionContext, repos: &[PathBuf], args: &[String]) -> Result<()> {
    if args.is_empty() {
        anyhow::bail!("No git command specified");
    }

    let formatter = PassthroughFormatter;

    run_parallel(
        ctx,
        repos,
        |repo| GitCommand::new(repo.clone(), args.to_vec()),
        &formatter,
    )
}
